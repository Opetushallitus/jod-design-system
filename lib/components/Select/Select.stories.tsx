import type { StoryObj } from '@storybook/react-vite';
import type { TitledMeta } from '../../utils';
import { Select } from './Select';

const meta = {
  title: 'Forms/Select',
  component: Select,
  tags: ['autodocs'],
} satisfies TitledMeta<typeof Select>;

export default meta;

type Story = StoryObj<typeof meta>;

const design = {
  type: 'figma',
  url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-7243&t=RDNOudQv5NR9oxZQ-4',
};

const options = [
  { label: 'Suomi', value: 'fi' },
  { label: 'Svenska', value: 'sv' },
  { label: 'English', value: 'en' },
];

const label = 'Choose language';

const decorators: Story['decorators'] = [
  (Story) => {
    return (
      <div className="ds:min-h-[250px]">
        <Story />
      </div>
    );
  },
];

export const Default: Story = {
  decorators,
  parameters: {
    design,
    docs: {
      description: {
        story: 'Select component for selecting a value from a list.',
      },
    },
  },
  args: {
    label,
    placeholder: label,
    options,
  },
};

export const Disabled: Story = {
  decorators,
  parameters: {
    design,
    docs: {
      description: {
        story: 'Disabled select component. User cannot interact with the component.',
      },
    },
  },
  args: {
    disabled: true,
    label,
    placeholder: label,
    options,
  },
};

export const WithPreselectedValue: Story = {
  decorators,
  parameters: {
    design,
    docs: {
      description: {
        story: 'Select component with preselected value',
      },
    },
  },
  args: {
    label,
    placeholder: label,
    selected: 'en',
    options,
  },
};
