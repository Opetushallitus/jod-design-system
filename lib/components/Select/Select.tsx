import { Listbox, ListboxButton, ListboxOption, ListboxOptions } from '@headlessui/react';
import React from 'react';
import { JodCaretDown, JodCaretUp } from '../../icons';
import { tidyClasses as tc } from '../../utils';

export interface SelectOptionsData<T extends string = string> {
  value: T;
  label: string;
}

interface SelectProps<T extends SelectOptionsData, U extends string = string> {
  /** Label for the component */
  label: string;
  /** Hide label. Still available for screenreaders */
  hideLabel?: boolean;
  /** Options for component */
  options: T[];
  /** Controlled mode */
  selected?: U;
  /** Callback on selection change */
  onChange?: (value: U) => void;
  /** Component is disabled for user interaction */
  disabled?: boolean;
  /** Placeholder text */
  placeholder: string;
  /** Classname to wrapper **/
  className?: string;
  /** Test id for querying in tests */
  dataTestId?: string;
}

export const Select = <U extends string = string, T extends SelectOptionsData<string> = SelectOptionsData<string>>({
  label,
  hideLabel = false,
  options,
  onChange: propOnChange,
  placeholder,
  selected,
  disabled = false,
  className = '',
  dataTestId,
}: SelectProps<T, U>) => {
  const inputId = React.useId();

  const [value, setValue] = React.useState<U | undefined>(selected);

  const onChange = (newValue: U) => {
    setValue(newValue);
    propOnChange?.(newValue);
  };

  React.useEffect(() => {
    setValue(selected);
  }, [selected]);

  const selectedOption = options.find((option) => (option.value as U) === value);

  return (
    <div className={tc(['ds:flex ds:flex-col ds:relative', className])} data-testid={dataTestId}>
      {!hideLabel && (
        <label htmlFor={inputId} className="ds:text-primary-gray ds:text-form-label ds:font-arial ds:mb-4">
          {label}
        </label>
      )}
      <div className="ds:flex ds:flex-row ds:relative">
        <Listbox onChange={onChange} disabled={disabled} value={value}>
          {({ open }) => (
            <div className="ds:flex ds:flex-row ds:w-full">
              <ListboxButton
                aria-label={label}
                className="ds:select-none ds:rounded ds:border ds:w-full ds:border-border-gray ds:bg-white ds:p-5 ds:text-primary-gray ds:disabled:text-inactive-gray ds:flex ds:justify-between ds:items-center"
                disabled={disabled}
                data-testid={dataTestId ? `${dataTestId}-button` : undefined}
              >
                {selectedOption ? (
                  <span className="ds:font-arial">{selectedOption.label}</span>
                ) : (
                  <span className="ds:font-arial ds:text-secondary-gray">{placeholder}</span>
                )}
                {open ? <JodCaretUp size={24} /> : <JodCaretDown size={24} />}
              </ListboxButton>
              <ListboxOptions
                modal={false}
                className="ds:bg-white ds:mt-3 ds:absolute ds:w-full ds:top-full ds:p-5 ds:m-0 ds:shadow-border ds:rounded-md ds:z-50 ds:empty:invisible"
                data-testid={dataTestId ? `${dataTestId}-options` : undefined}
              >
                {options.map((option) => (
                  <ListboxOption
                    key={option.value}
                    className="ds:py-3 ds:text-heading-4 ds:ml-5 ds:text-primary-gray ds:cursor-pointer ds:data-focus:underline ds:data-focus:text-accent ds:hover:underline ds:hover:text-accent"
                    value={option.value}
                  >
                    {option.label}
                  </ListboxOption>
                ))}
              </ListboxOptions>
            </div>
          )}
        </Listbox>
      </div>
    </div>
  );
};
