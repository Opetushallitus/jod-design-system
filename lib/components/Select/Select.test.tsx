import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';
import { Select } from './Select';

describe('Select', () => {
  const options = [
    { value: 'fi', label: 'Suomi' },
    { value: 'sv', label: 'Svenska' },
    { value: 'en', label: 'English' },
  ];

  const placeholder = 'placeholder text';

  describe('Snapshot testing', () => {
    it('should render with defaults', () => {
      const { container } = render(<Select label="Label" options={options} placeholder={placeholder} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should not show label', () => {
      const { container } = render(<Select hideLabel label="Label" options={options} placeholder={placeholder} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render as disabled', () => {
      const { container } = render(<Select disabled label="Label" options={options} placeholder={placeholder} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('selected', () => {
    it('should show placeholder', async () => {
      render(<Select label="Label" options={options} placeholder={placeholder} />);
      const select: HTMLButtonElement = screen.getByRole('button');
      expect(select.textContent).toBe(placeholder);
      const user = userEvent.setup();
      await user.click(select);
      expect(screen.getByRole('option', { name: 'Suomi' }).ariaSelected).toBe('false');
      expect(screen.getByRole('option', { name: 'Svenska' }).ariaSelected).toBe('false');
      expect(screen.getByRole('option', { name: 'English' }).ariaSelected).toBe('false');
    });

    it('should select the wanted one', async () => {
      render(<Select label="Label" options={options} placeholder={placeholder} selected="sv" />);
      const select: HTMLButtonElement = screen.getByRole('button');
      expect(select.textContent).toBe('Svenska');
      const user = userEvent.setup();
      await user.click(select);
      expect(screen.getByRole('option', { name: 'Svenska' }).ariaSelected).toBe('true');
    });

    it('should still allow user to another', async () => {
      render(<Select label="Label" options={options} placeholder={placeholder} selected="fi" />);
      const select: HTMLButtonElement = screen.getByRole('button');
      // Click the expand button and select English
      const user = userEvent.setup();
      await user.click(select);
      await user.selectOptions(screen.getByRole('listbox'), 'English');

      // Assert that the value is English, click the expand button and check that English is selected
      expect(select.textContent).toBe('English');
      await user.click(select);
      expect(screen.getByRole('option', { name: 'English' }).ariaSelected).toBe('true');
    });
  });

  describe('onChange', () => {
    it('is called on user selection', async () => {
      const mockCallback = vi.fn();
      const user = userEvent.setup();

      render(<Select label="Label" options={options} placeholder={placeholder} onChange={mockCallback} />);
      await user.click(screen.getByRole('button'));
      await user.selectOptions(screen.getByRole('listbox'), 'English');
      expect(mockCallback).toHaveBeenCalledTimes(1);
    });
  });

  describe('label', () => {
    it('should render with label', () => {
      render(<Select label="Label" options={[]} placeholder={placeholder} dataTestId="cb" />);
      expect(screen.getAllByText('Label')).not.toBeNull();
      expect(screen.getByTestId('cb')).toBeInTheDocument();
    });
  });

  describe('hideLabel', () => {
    it('should add aria-label when label is hidden', () => {
      render(<Select label="Non-visible label" options={[]} hideLabel={true} placeholder={placeholder} />);
      expect(screen.getByRole('button', { name: 'Non-visible label' })).not.toBeNull();
    });
  });
});
