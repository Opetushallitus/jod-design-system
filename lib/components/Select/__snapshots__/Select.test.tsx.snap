// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select > Snapshot testing > should not show label 1`] = `
<div
  class="ds:flex ds:flex-col ds:relative"
>
  <div
    class="ds:flex ds:flex-row ds:relative"
  >
    <div
      class="ds:flex ds:flex-row ds:w-full"
      data-headlessui-state=""
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-label="Label"
        class="ds:select-none ds:rounded ds:border ds:w-full ds:border-border-gray ds:bg-white ds:p-5 ds:text-primary-gray ds:disabled:text-inactive-gray ds:flex ds:justify-between ds:items-center"
        data-headlessui-state=""
        id="headlessui-listbox-button-«ra»"
        type="button"
      >
        <span
          class="ds:font-arial ds:text-secondary-gray"
        >
          placeholder text
        </span>
        <svg
          fill="currentColor"
          height="24"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 15.6964L6 9.69639L7.4 8.29639L12 12.8964L16.6 8.29639L18 9.69639L12 15.6964Z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Select > Snapshot testing > should render as disabled 1`] = `
<div
  class="ds:flex ds:flex-col ds:relative"
>
  <label
    class="ds:text-primary-gray ds:text-form-label ds:font-arial ds:mb-4"
    for="«re»"
  >
    Label
  </label>
  <div
    class="ds:flex ds:flex-row ds:relative"
  >
    <div
      class="ds:flex ds:flex-row ds:w-full"
      data-disabled=""
      data-headlessui-state="disabled"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-label="Label"
        class="ds:select-none ds:rounded ds:border ds:w-full ds:border-border-gray ds:bg-white ds:p-5 ds:text-primary-gray ds:disabled:text-inactive-gray ds:flex ds:justify-between ds:items-center"
        data-disabled=""
        data-headlessui-state="disabled"
        disabled=""
        id="headlessui-listbox-button-«rh»"
        type="button"
      >
        <span
          class="ds:font-arial ds:text-secondary-gray"
        >
          placeholder text
        </span>
        <svg
          fill="currentColor"
          height="24"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 15.6964L6 9.69639L7.4 8.29639L12 12.8964L16.6 8.29639L18 9.69639L12 15.6964Z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Select > Snapshot testing > should render with defaults 1`] = `
<div
  class="ds:flex ds:flex-col ds:relative"
>
  <label
    class="ds:text-primary-gray ds:text-form-label ds:font-arial ds:mb-4"
    for="«r0»"
  >
    Label
  </label>
  <div
    class="ds:flex ds:flex-row ds:relative"
  >
    <div
      class="ds:flex ds:flex-row ds:w-full"
      data-headlessui-state=""
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-label="Label"
        class="ds:select-none ds:rounded ds:border ds:w-full ds:border-border-gray ds:bg-white ds:p-5 ds:text-primary-gray ds:disabled:text-inactive-gray ds:flex ds:justify-between ds:items-center"
        data-headlessui-state=""
        id="headlessui-listbox-button-«r3»"
        type="button"
      >
        <span
          class="ds:font-arial ds:text-secondary-gray"
        >
          placeholder text
        </span>
        <svg
          fill="currentColor"
          height="24"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 15.6964L6 9.69639L7.4 8.29639L12 12.8964L16.6 8.29639L18 9.69639L12 15.6964Z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
